staging/src/k8s.io/apimachinery/pkg/conversion/conversiontest/apis/v2/types.go

```diff
@@ -0,0 +1,35 @@
+package v2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// To convert from v1 to v2, you have to either default the Type or write logic in your conversion method that effectively
+// does the defaulting to decide what to do.  If you're migrating an older codebase, you need to make sure that nothing
+// in the conversion path relies on having required fields set.  We used to simply allow a failure if that were the case,
+// but if you need to tolerate it for APPLY you have code to inspect.
+// In addition, you need to preserve the pointer aspect of Type through the internal hub to inform the internal->v2 conversion, which didn't use to be the case since defaulting used to happen when going v1->internal.
```

Comment:
The potential bad thing is:

1. after a v2 -> v1 downgrade, the Type field is now set, when it wasn't before. (IMO it is an API bug that this happens, but API bugs must be lived with.)
2. The user re-applies their original manifest, still in the v1 format, which never had the Type field set.
3. The service observes that the type field is deleted and deletes it.
4. The defaulting path adds it back.

So, in this case, I think nothing bad happens.

One of your points is that step 1 is broken, as we (correctly) relied on defaulting to set this and not the conversion path. Response to that below.